modules/transaction-payment/src/lib.rs

```diff
@@ -611,59 +673,206 @@ where
-					Some(target_currency_id) if *target_currency_id == native_currency_id => {
-						let supply_currency_id = *trading_path.first().expect("these's first guaranteed by match");
-						// calculate the supply limit according to oracle price and the slippage limit,
-						// if oracle price is not avalible, do not limit
-						let max_supply_limit = if let Some(target_price) =
-							T::PriceSource::get_relative_price(*target_currency_id, supply_currency_id)
-						{
-							Ratio::one()
-								.saturating_sub(T::MaxSwapSlippageCompareToOracle::get())
-								.reciprocal()
-								.unwrap_or_else(Ratio::max_value)
-								.saturating_mul_int(target_price.saturating_mul_int(amount))
-						} else {
-							PalletBalanceOf::<T>::max_value()
-						};
-
-						if T::DEX::swap_with_exact_target(
-							who,
-							&trading_path,
-							amount.unique_saturated_into(),
-							<T as Config>::MultiCurrency::free_balance(supply_currency_id, who)
-								.min(max_supply_limit.unique_saturated_into()),
-						)
-						.is_ok()
+		// make sure add extra gap to keep alive after swap.
+		let amount = fee.saturating_add(native_existential_deposit.saturating_sub(total_native));
+		// native is not enough, try swap native from treasury fee pool to pay fee and gap.
+		Self::swap_native_asset(who, amount);
+	}
+
+	/// Iterate order list, break if can swap out enough native asset amount with user's foreign
+	/// asset. make sure trading path is exist in dex, if the trading pair is not exist in dex, even
+	/// though we have setup it in charge fee pool, we can't charge fee with this foreign asset.
+	fn swap_native_asset(who: &T::AccountId, amount: Balance) {
+		let native_currency_id = T::NativeCurrencyId::get();
+		for trading_path in Self::get_trading_path(who) {
+			if let Some(target_currency_id) = trading_path.last() {
+				if *target_currency_id == native_currency_id {
+					let supply_currency_id = *trading_path.first().expect("should match a non native asset");
+					if Self::swap_from_treasury_or_dex(who, amount, supply_currency_id).is_ok() {
+						break;
+					}
+				}
+			}
+		}
+	}
+
+	/// swap treasury account's native asset with user's foreign asset from treasury fee pool when
+	/// treasury account's native asset balance is gt threshold. or-else, swap treasury account's
+	/// native asset with treasury account's foreign asset from dex swap liquid pool.
+	#[transactional]
+	fn swap_from_treasury_or_dex(
+		who: &T::AccountId,
+		amount: Balance,
+		supply_currency_id: CurrencyId,
+	) -> sp_std::result::Result<(), DispatchError> {
+		match TokenFixedRate::<T>::get(supply_currency_id) {
+			None => {
+				// if the token rate is none(i.e. disable by council or on purpose), return err.
+				Err(DispatchError::CannotLookup)
+			}
+			Some(mut rate) => {
+				// user's supply asset should have enough balance based on old fix rate.
+				ensure!(
+					T::MultiCurrency::free_balance(supply_currency_id, who) >= rate.saturating_mul_int(amount),
```

Comment:
no need to check free balance, just do the transfer here, which will check the free balance.